src/common/m_boundary_common.fpp

```diff
@@ -0,0 +1,582 @@
+                        ${COLOR_FUNC_BC("k,l,-j","k,l,p - (j-1)")}$
+                    elseif (bc_type(3, -1)%sf(k, l, 0) == -2) then
+                        ${COLOR_FUNC_SLIP_WALL_BC(3,"k,l,-j","k,l,j-1")}$
+                    else
+                        ${COLOR_FUNC_BC("k,l,-j","k,l,0")}$
+                    end if
+                end do
+            end do
+        end if
+
+        if (bcze >= 0) then
+            call s_mpi_sendrecv_capilary_variables_buffers(c_divs, 3, 1)
+        else
+            !$acc parallel loop collapse(2) gang vector default(present)
+            do l = -buff_size, n + buff_size
+                do k = -buff_size, m + buff_size
+                    if (bc_type(3, 1)%sf(k, l, 0) == -1) then
+                        ${COLOR_FUNC_BC("k,l,p+j","k,l,j-1")}$
+                    elseif (bc_type(3, 1)%sf(k, l, 0) == -2) then
+                        ${COLOR_FUNC_SLIP_WALL_BC(3,"k,l,p+j","k,l,p - (j-1)")}$
+                    else
+                        ${COLOR_FUNC_BC("k,l,p+j","k,l,p")}$
+                    end if
+                end do
+            end do
+        end if
+
+    end subroutine s_populate_capillary_buffers
+#endif
+
+    subroutine s_create_mpi_types(bc_type)
+
+        type(integer_field), dimension(1:num_dims, -1:1) :: bc_type
+
+#ifdef MFC_MPI
+        integer :: dir, loc
+        integer, dimension(3) :: sf_start_idx, sf_extents_loc
+        integer :: ifile, ierr, data_size
+
+        do dir = 1, num_dims
+            do loc = -1, 1, 2
+                sf_start_idx = (/0, 0, 0/)
+                sf_extents_loc = shape(bc_type(dir, loc)%sf)
+
+                call MPI_TYPE_CREATE_SUBARRAY(num_dims, sf_extents_loc, sf_extents_loc, sf_start_idx, &
+                                              MPI_ORDER_FORTRAN, MPI_INTEGER, MPI_BC_TYPE_TYPE(dir, loc), ierr)
+                call MPI_TYPE_COMMIT(MPI_BC_TYPE_TYPE(dir, loc), ierr)
+            end do
+        end do
+
+        do dir = 1, num_dims
+            do loc = -1, 1, 2
+                sf_start_idx = (/0, 0, 0/)
+                sf_extents_loc = shape(bc_buffers(dir, loc)%sf)
+
+                call MPI_TYPE_CREATE_SUBARRAY(num_dims, sf_extents_loc, sf_extents_loc, sf_start_idx, &
+                                              MPI_ORDER_FORTRAN, mpi_p, MPI_BC_BUFFER_TYPE(dir, loc), ierr)
+                call MPI_TYPE_COMMIT(MPI_BC_BUFFER_TYPE(dir, loc), ierr)
+            end do
+        end do
+#endif
+    end subroutine s_create_mpi_types
+
+    subroutine s_finalize_boundary_common_module()
+
+        deallocate (bc_buffers(1, -1)%sf)
+        deallocate (bc_buffers(1, 1)%sf)
+        if (n > 0) then
+            deallocate (bc_buffers(2, -1)%sf)
+            deallocate (bc_buffers(2, 1)%sf)
+            if (p > 0) then
+                deallocate (bc_buffers(3, -1)%sf)
+                deallocate (bc_buffers(3, 1)%sf)
+            end if
+        end if
+
+        deallocate (bc_buffers)
```

Comment:
Replace `deallocate` with `@:DEALLOCATE` and only deallocate if needed (see earlier comment about allocating).